[DX-815] Plugins Overview Refactor (#5330)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-gateway/advanced-configurations/plugins/api-config/overview.md

---
date: 2024-08-28
title: Overview
description: "This section explains an overview for how to configure plugins for APIs"
tags: [ "plugins" ]
---

This page provides an overview for configuring one or more API-level custom plugins that can be optionally setup to execute for each of the following [hooks]({{< ref "plugins/plugin-types/plugintypes#plugin-and-hook-types" >}}) in the API request/response lifecycle: *Pre*, *Authentication*, *Post*, *Post Authentication* and *Response*. If you wish to learn how to configure API-level Analytics plugins please refer to the [Analytics plugins]({{< ref "plugins/plugin-types/analytics-plugins" >}}) page. 

If you desire fine grained control at the endpoint level then it is also possible to configure [per-endpoint plugins]({{< ref "product-stack/tyk-gateway/middleware/endpoint-plugin" >}}). These are custom Golang plugins that are triggered at the end of the request processing chain before API-level *Post* plugins are executed.

---

## Introduction

> There are three scenarios for configuring Pre, Authentication, Post Authentication and Response plugins for an API:

Reviewer: These aren't scenarios - these are locations where Tyk Gateway can find the plugin function. What about `Post` plugins? Are these deliberately missing? I'm not sure you need to list the different hooks here.

Author: Updated to with _There are three locations where Tyk Gateway can find plugin functions:_

1. **gRPC plugins**: Plugins are implemented by a gRPC server with the associated configuration specified with the API definition. For further details on how to configure gRPC plugins, please refer to our [gRPC]({{< ref "plugins/supported-languages/rich-plugins/grpc" >}}) documentation.

> 2. **Plugin bundles**: The plugin source code and configuration is bundled into a zip file that is served by a remote web server. For further details and how to configure, checkout the [plugin bundles]({{< ref "plugins/how-to-serve-plugins/plugin-bundles" >}}) page.

Reviewer: ```suggestion
2. **Plugin bundles**: The plugin source code and configuration are bundled into a zip file that is served by a remote web server. For further details see the [plugin bundles]({{< ref "plugins/how-to-serve-plugins/plugin-bundles" >}}) page.
```

3. **Local plugins**: Plugins are implemented by functions within source code files located on the Gateway's file system. The API Definition allows the source code file path and function name to be configured for each plugin. For further details read on.

> ## Plugin configuration

Reviewer: Typically in these overview pages we cover the concepts rather than the detail of the configuration (that's then in the Tyk OAS/Tyk Classic pages to ensure clear scope of terminology).

So I would remove this section but add:
- a section that covers how Tyk Gateway finds/identifies the correct code to run (path/function name/driver (note that the "driver" for JavaScript is "Otto"))
- another section that covers what data is actually passed to the plugin function (raw body/session state)

You've already got a section (below) for language that mentions "driver".

Author: Thanks Andy, I have raised a [ticket](https://tyktech.atlassian.net/browse/DX-1657) to cover in separate iteration

@letzya

Each plugin for an API can be configured within the API Definition with the following details:

| Property | Description |
|-------|-------------|
| `Enabled` | When true, the plugin is activated |
| `Name` | A name used to identify the plugin |
| `Path` | The path to the source code file on the Tyk Gateway file system |
| `Function name` | The name of the function that implements the plugin. The function should exist within the source code file referenced in `path` |
| `Raw body only` | When set to true, this flag indicates that only the raw request body should be processed |
| `Require session state`| When set to true, Tyk Gateway will serialize the request session state and pass it as an argument to the function that implements the plugin in the target language. This is applicable to Post, Response, and Authentication hooks only |

---

## Language configuration

> All plugins for an API should be implemented using **one** of the [supported target languages]({{< ref "plugins/supported-languages" >}}). We configure the plugin implementation language for an API by specifying a [plugin driver]({{< ref "plugins/supported-languages#plugin-driver-names" >}}).

Reviewer: I think this is only for Local plugins and Bundled plugins - gRPC can use any language.

The "driver" isn't just used for language, as it also indicates if it's gRPC - so I suggest a little more detail here on the driver options (this [link](https://tyk.io/docs/plugins/supported-languages/#plugin-driver-names) isn't descriptive and really only provides a reference table)

Author: Updated with:

_For local and bundle plugins a [plugin driver]({{< ref "plugins/supported-languages#plugin-driver-names" >}}) is configured to specify the plugin implementation language. If using gRPC plugins a `grpc` plugin driver should be used to instruct Tyk to request execution of plugins from within a gRPC server that is external to the Tyk process. This offers additional language support since Tyk can integrate with a gRPC server that is implemented using any supported [gRPC language](https://grpc.io/docs/)._
